internal/operator-controller/authorization/rbac.go

```diff
@@ -358,134 +364,116 @@ type escalationChecker struct {
-		ctx = request.WithRequestInfo(ctx, &request.RequestInfo{APIGroup: rbacv1.GroupName, Resource: "rolebindings", IsResourceRequest: true})
-		return ec.checkRoleBindingEscalation(ctx, v)
 	case *rbacv1.ClusterRole:
-		ctx = request.WithRequestInfo(ctx, &request.RequestInfo{APIGroup: rbacv1.GroupName, Resource: "clusterroles", IsResourceRequest: true})
-		return ec.checkClusterRoleEscalation(ctx, v)
-	case *rbacv1.ClusterRoleBinding:
-		ctx = request.WithRequestInfo(ctx, &request.RequestInfo{APIGroup: rbacv1.GroupName, Resource: "clusterrolebindings", IsResourceRequest: true})
-		return ec.checkClusterRoleBindingEscalation(ctx, v)
-	default:
-		return fmt.Errorf("unknown object type %T", v)
-	}
-}
-
-func (ec *escalationChecker) checkClusterRoleEscalation(ctx context.Context, clusterRole *rbacv1.ClusterRole) error {
-	if rbacregistry.EscalationAllowed(ctx) || rbacregistry.RoleEscalationAuthorized(ctx, ec.authorizer) {
-		return nil
-	}
-
-	// to set the aggregation rule, since it can gather anything, requires * on *.*
-	if hasAggregationRule(clusterRole) {
-		if err := validation.ConfirmNoEscalation(ctx, ec.ruleResolver, fullAuthority); err != nil {
-			return fmt.Errorf("must have cluster-admin privileges to use an aggregationRule: %w", err)
+		// Merge extra ClusterRole rules if present.
+		key := types.NamespacedName{Name: v.Name}
+		if extra, ok := ec.extraClusterRoles[key]; ok {
+			v.Rules = append(v.Rules, extra.Rules...)
 		}
-	}
+		// Convert external ClusterRole to internal representation.
+		var internalClusterRole rbacinternal.ClusterRole
+		if err := rbacv1helpers.Convert_v1_ClusterRole_To_rbac_ClusterRole(v, &internalClusterRole, nil); err != nil {
+			return err
+		}
+		storage := policybased.NewStorage(&FakeStandardStorage{}, ec.authorizer, ec.ruleResolver)
+		_, err := storage.Create(ctx, &internalClusterRole, noOpValidate, opts)
+		return err
 
-	if err := validation.ConfirmNoEscalation(ctx, ec.ruleResolver, clusterRole.Rules); err != nil {
+	case *rbacv1.ClusterRoleBinding:
+		// Convert external ClusterRoleBinding to internal representation.
+		var internalClusterRoleBinding rbacinternal.ClusterRoleBinding
+		if err := rbacv1helpers.Convert_v1_ClusterRoleBinding_To_rbac_ClusterRoleBinding(v, &internalClusterRoleBinding, nil); err != nil {
+			return err
+		}
+		storage := policybasedClusterRoleBinding.NewStorage(&FakeStandardStorage{}, ec.authorizer, ec.ruleResolver)
+		_, err := storage.Create(ctx, &internalClusterRoleBinding, noOpValidate, opts)
 		return err
-	}
-	return nil
-}
 
-func (ec *escalationChecker) checkClusterRoleBindingEscalation(ctx context.Context, clusterRoleBinding *rbacv1.ClusterRoleBinding) error {
-	if rbacregistry.EscalationAllowed(ctx) {
-		return nil
-	}
+	case *rbacv1.Role:
+		// Merge extra Role rules if present.
+		key := types.NamespacedName{Namespace: v.Namespace, Name: v.Name}
+		if extra, ok := ec.extraRoles[key]; ok {
+			v.Rules = append(v.Rules, extra.Rules...)
+		}
```

Comment:
This is not the right place to merge extra rules. The `extraRoles` and `extraClusterRoles` need to be used to augment rules returned by the rules resolver when evaluating a role binding or cluster role binding.

The rules resolver only knows about roles and cluster roles that exist in our cache (via KAS and etcd). However, we may have role bindings and cluster role bindings in our manifest that reference other roles and cluster roles in our manifest.

So when `ConfirmNoEscalation` eventually executes, the rules passed to it need to include the rules that _would_ be bound if the referenced roles or cluster roles _did_ exist.

Reply:
@bentito likewise confirming this is also gone?

Reply:
yep gone marking resolved